### DISCUSSION TITLE
Encounter error on "error: undefined symbol: $GL" with Strict=1

### DISCUSSION THREAD

**Discussion by shinji-TH** [emscripten-core/emscripten#19926] [Q&A] [answered]
Hi, I have a small app compile in with the following flags 

> /home/yapt/emsdk/upstream/emscripten/emcc.py /home/yapt/EmscriptenTestProjSkeleton/main.cpp -g -I test_file.h -o /home/yapt/EmscriptenTestProjSkeleton/build/index.html -O0 --js-library jslibrary.js --pre-js preload.js --post-js postload.js --profiling --profiling-funcs -s ASSERTIONS=1 -s SAFE_HEAP=0 -s DEMANGLE_SUPPORT=1 -lnodefs.js -std=c++20 -D_LIBCPP_ENABLE_CXX17_REMOVED_AUTO_PTR -s INCLUDE_FULL_LIBRARY=1 -fstrict-aliasing -fPIC -s WASM_BIGINT=1 -pthread -s PTHREAD_POOL_SIZE=2 -s PROXY_TO_PTHREAD -s FETCH=1 -s FETCH_SUPPORT_INDEXEDDB=0 -Wno-deprecated-builtins -s WASM=1 -s ALLOW_MEMORY_GROWTH=1 -s EVAL_CTORS=0 -s MAIN_MODULE=2 -s AUTOLOAD_DYLIBS=0 -s NO_EXIT_RUNTIME=0 -s INLINING_LIMIT=1 -s FORCE_FILESYSTEM=1 --emit-symbol-map -s LEGACY_GL_EMULATION=0 -s INITIAL_MEMORY=1000MB -s ENVIRONMENT=web,worker -s USE_SDL=0 -s STRICT=1 -s ALLOW_UNIMPLEMENTED_SYSCALLS=1 -s SPLIT_MODULE=1 -s USE_ICU=1 -s EXPORTED_RUNTIME_METHODS=[allocate,writeArrayToMemory] 

With the flag `-s ERROR_ON_UNDEFINED_SYMBOLS=1`
i would result in the following errors

> wasm-ld: error: symbol exported via --export not found: glGetString

Without the flag would result in the following warnings
>warning: undefined symbol: $GL (referenced by $GLFW__deps: ['emscripten_get_now','$GL','$Browser','$GLFW_Window','$allocateUTF8','$FS','$GLFW','$getWasmTableEntry'], referenced by $GLFW_Window__deps: ['$GLFW'], referenced by top-level compiled C/C++ code)
warning: undefined symbol: glGetString (referenced by glfwExtensionSupported__deps: ['glGetString','$GLFW'], referenced by top-level compiled C/C++ code)

I did some tracing and understand the errors came from the following changes: https://github.com/emscripten-core/emscripten/pull/19057

From the PR, it seem GLFW should be included unless the flag `MINIMAL_RUNTIME` is set as zero. 
However its dependency `$GL`, which was define in `library_webgl.js`, was only included when `AUTO_JS_LIBRARIES`  is set as 1 or `MAX_WEBGL_VERSION` is greater than 2. With `STRICT=1`, the libraries are automatically excluded and resulting in the missing symbols. 

Hence wish to check with @sbc100 about it.

**Comment by sbc100** [comment, marked as answer]
The is working as intended.  In `STRICT` mode (i.e. with `AUTO_JS_LIBRARIES=0`) you need to explicitly link any libraries that you use.  

In this case I think you can just add `-lGL` (or `-lGLESv2`) to your command line).

**Reply by shinji-TH** [reply]
Noted, just finding it odd to add in `-lGL` when my app has no usage on any WebGL functionalities.
Thanks!!

**Reply by sbc100** [reply]
Oh I see... it looks like the GL dependency is due to GLFW.  Do you know where the GLFW dependency comes from?

**Reply by sbc100** [reply]
Oh, I guess the `-sINCLUDE_FULL_LIBRARY` setting might be the issue.  Is there some reason you are using that option?

**Reply by sbc100** [reply]
Ah yes I can repro with just that setting:

```
$ emcc -sSTRICT -sINCLUDE_FULL_LIBRARY test/hello_world.c
error: undefined symbol: $GL (referenced by $GLFW, referenced by $GLFW_Window, referenced by root reference (e.g. compiled C/C++ code))
warning: To disable errors for undefined symbols use `-sERROR_ON_UNDEFINED_SYMBOLS=0`
warning: GL may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: glGetString (referenced by glfwExtensionSupported, referenced by root reference (e.g. compiled C/C++ code))
warning: _glGetString may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
Error: Aborting compilation due to previous errors
emcc: error: '/usr/local/google/home/sbc/bin/node-v20.0.0-v8-canary202302081604228b65-linux-x64/bin/node --experimental-wasm-memory64 /usr/local/google/home/sbc/dev/wasm/emscripten/src/compiler.js /tmp/tmp2i5y0reu.json' failed (returned 1)
```

**Reply by shinji-TH** [reply]
I believe the flag `-sINCLUDE_FULL_LIBRARY` was included due to the app is required to load side modules via dlopen....

**Comment by sbc100** [comment]
Fixed in #19935

**Reply by shinji-TH** [reply]
Thanks!!
